File: C++ Projects/Quiz/quiz.cpp
//-------------------------
//class implentation(.cpp)
//author: Noah Funderburgh
//File: quiz.cpp
//-------------------------
#include "quiz.h"
#include "questiontf.h"
#include "questionmc.h"
#include "questionsa.h"
using namespace std;
//-------------------------
quiz::quiz(){

}
//-------------------------
bool quiz::isAnswerMulti(string line){
    int countSemiColons = 0;
    if(line[0] == 'M'){
        for(int i = 0;i<=line.length();i++){
            if(line[i] == ':'){
                if(line[i+1] == ':'){
                    cerr << "No given answer between Colons, ";
                    return false;
                }
                countSemiColons++;
            }
        }
        countSemiColons++;

        char isLower = countSemiColons + 'a';
        char isUpper = countSemiColons + 'A';
        int size = line.size();
        if(line[size-1] >= 'a' && line[size-1] < isLower){
            return true;
        }
        if(line[size-1] >= 'A' && line[size-1] < isUpper){
            return true;
        }
        else{
            cerr << "Answer does not match questions, ";
            return false;
        }
    }
    else{ 
        return true;
    }
}
//-------------------------
bool quiz::isType(string line){
    if( line[0] != 'M' && line[0]!='S' && line[0] != 'T'){
        cerr << "Type is inncorrect, ";
        return false;
    }
    return true;
}
//-------------------------
bool quiz::spacedCorrect(string line){
    int countMultiQuestions = 0;
    int countSpacers = 0;
    if(line[1] != '|' || line[3] != '|'){
        cerr << "Delimiters are in inncorrect places, ";
        return false;
    }
    for(int i = 0;i<=line.length();i++){
        if(line[0] == 'T' || line[0] == 'S'){
            if(line[i] == ':'){
                cerr << "There shouldnt be Colons, ";
                return false;
            }
        }
        if(line[i] == ':'){
            countMultiQuestions++;
        }
        if(line[i] == '|'){
            countSpacers++;
        }
    }
    if(line[0] == 'M'){
        if(countSpacers != 4){
            cerr << "There should only be 4 delimiters, ";
            return false;
        }
    }
    if(line[0] == 'T' || line[0] == 'S'){
        if(countSpacers != 3){
            cerr << "There should only be 3 delimiters, ";
            return false;
        }
    }
    return true;
}
//-------------------------
bool quiz::isLevel(string line){
    if(line[2] == '0'){ // 1-9 is our level range
        cerr << "Level is inncorrect, ";
        return false;
    }
    return true;
}
//-------------------------
bool quiz::checkLine(string line){
    int correct = 0;

    if(isAnswerMulti(line) == true)
        correct++;
    if(isType(line) == true)
        correct++;
    if(spacedCorrect(line) == true)
        correct++;
    if(isLevel(line) == true)
        correct++;
    if(correct == 4)
        return true;
    else
        return false;
}
//-------------------------
bool quiz::isComment(string line){
    if(line[0] == '#')
        return false;
    else
        return true;
}
//-------------------------
bool quiz::loadQuestions(string filename){
    ifstream infile(filename.c_str());
    if(infile.is_open()){
        int lineCount = 1;
        string line;
        question* temp = NULL;
        questionTF questTF;
        questionMC questMC;
        questionSA questSA;
        while(getline(infile,line)){

            stringstream ss(line);
            string token;
            if(isComment(line) == true){
                if(checkLine(line) == true){
                    getline(ss,token,'|');//type
                    char letter = token.at(0); 

                    if(letter == 'S')
                        temp = new questionSA;                      
                    if(letter == 'T')
                        temp = new questionTF;
                    if(letter == 'M')
                        temp = new questionMC;

                    getline(ss,token,'|');
                    int level = atoi(token.c_str());

                    if(letter == 'T' || letter == 'S'){
                        getline(ss,token,'|');
                        temp->setQuestion(token);
                    }
                    if(letter == 'M'){
                        getline(ss,token,'|');
                        string questionToken = token;
                        getline(ss,token,'|');
                        string questionPlusChoices = questionToken + ':' + token;
                        temp->setQuestion(questionPlusChoices);
                    }
                    getline(ss,token,'|');//Answer
                    temp->setAnswer(token);

                    questions.push_back(temp);
                }
                else{
                    cerr << "on Line " << lineCount << " question will be removed." << endl;
                }
            }
            lineCount++;
        }
        infile.close();
        return true;
    }
    else{
        cout << "cannot open the file" << endl;
        return false;
    }
}
//-------------------------
void quiz::dumpQuestions(){
    int size = questions.size();
    for(int i = 0; i<size;i++){
        questions.at(i)->showQuestion();
        cout << "Answer: ";
        questions.at(i)->showAnswer();
        cout <<"---------------------" << endl;
    }
}
//-------------------------
int quiz::deliverQuiz(){
    int correct = 0;
    int size =  questions.size();
    for(int i = 0; i<size;i++){
        questions.at(i)->showQuestion();

        string userGuess;
        cout <<"> ";
        cin >> userGuess;

        if(questions[i]->checkAnswer(userGuess)){
            cout << "Correct! Good Job!" << endl << endl;
            questions.at(i)->markCorrect();
            correct++;
        }
        else{ 
            cout << "Sorry, the answer was ";
            questions.at(i)->showAnswer();
            cout << endl;
        }
    }
    return correct;
}
//-------------------------
int quiz::numQuestions(){
    int size =  questions.size();
    return size;
}
//-------------------------

File: Java Projects/Quiz/Quiz.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.Vector;

/**
 * Implements steps to setup the quiz and give the user the quiz along with
 * telling the user whether there right or wrong. It also allows them to re take
 * it up to 100%
 *
 * @author Noah Funderburgh
 * @version 1.0
 * @since 3/3/2020
 */

public class Quiz {
	/**
	 * This is main which runs the quiz.
	 * 
	 * @param args - should be text file and second argument should be -d if you
	 *             would like a dump of the questions & answers.
	 */
	public static void main(String[] args) {

		String file = args[0];
		String dump = "-d";
		if (args.length == 2) {
			if (args[1].contentEquals(dump)) {
				loadQuestions(args[0]);
				dumpQuestions();
			} else {
				System.err.print("Failed To dump, Please use format (excuatable -d   )");
			}
		} else {
			loadQuestions(args[0]);
			System.out.println("Hope your read. Its quiz Time!!");
			System.out.println("After each answer is displayed, press enter to see the next question");

			if (args.length == 2) {
				if (args[1] == dump) {
					loadQuestions(args[0]);
					dumpQuestions();
				}
			}
			double numCorrect = deliverQuiz();
			double numQuestions = numQuestions();
			double percent = (numCorrect / numQuestions) * 100;
			System.out.println("You got " + numCorrect + " out of " + numQuestions + " which is " + percent + "%");
		}
	}

//-------------------------------------------------------
	/**
	 * The vector in which questions for our quiz our stored
	 */
	static Vector<Question> questions = new Vector<Question>(10);
//-------------------------------------------------------
	/**
	 * This public int takes the total number of questions that contain errors and
	 * sends its to deliver quiz so it knows that we how many questions that have no
	 * errors.
	 */
	static int totalUnErrorQuestions = 0;

//-------------------------------------------------------
	/**
	 * Loads all questions in given txt file into vector and Validates text file
	 * before loading them to check for errors.
	 * 
	 * @param args This is the text file you want to load.
	 * @return bool This returns true is text file was loaded properly else returns
	 *         false.
	 */
	public static boolean loadQuestions(String args) {
		int lineNum = 0;
		File file = new File(args);
		try {
			Scanner fsc = new Scanner(file);
			while (fsc.hasNextLine()) {
				lineNum++;
				String line = fsc.nextLine();
				Question temp = new Questiontf();
				char letter1 = line.charAt(0);
				char letter2 = line.charAt(1);

				if (Validate.isValid(line, lineNum)) {
					totalUnErrorQuestions++;
					if (letter1 == 'S' && letter2 == 'A')
						temp = new Questionsa();
					if (letter1 == 'T' && letter2 == 'F')
						temp = new Questiontf();
					if (letter1 == 'M' && letter2 == 'C' || letter1 == 'M' && letter2 == 'A')
						temp = new Questionmc();
					String[] tokens = line.split("\\|");
					temp.setType(tokens[0]);
					temp.setLevel(tokens[1]);
					if (line.charAt(0) == 'M' && line.charAt(1) == 'C'
							|| line.charAt(0) == 'M' && line.charAt(1) == 'A') {
						String question = tokens[2] + ":" + tokens[3];
						temp.setQuestion(question);
						temp.setAnswer(tokens[4]);
					} else {
						temp.setQuestion(tokens[2]);
						temp.setAnswer(tokens[3]);
					}
					questions.add(temp);
				}
			}
			fsc.close();
		} catch (FileNotFoundException e) {
			System.err.println("Error: " + file.getAbsolutePath() + " not found.");
		}
		return true;
	}

//-------------------------------------------------------
	/**
	 * Dumps all questions and answers onto interface for a quick way to make sure
	 * it is reading your text file correctly.
	 */
	static void dumpQuestions() {
		int size = questions.size();
		for (int i = 0; i < size; i++) {
			System.out.println("----------------------------------");
			questions.get(i).showQuestion();
			System.out.print("The answer is: ");
			questions.get(i).showAnswer();

		}
	}

//-------------------------------------------------------	
	/**
	 * Delivers quiz to user. Once all questions are answered user can retake
	 * Incorrect answers or end the quiz. This can continue to happen Until all
	 * answers are correct or until user gives up and wants to end the quiz.
	 * 
	 * @return int Returns the num of correct questions answered during the quiz.
	 */
	static int deliverQuiz() {
		int correct = 0;
		int size = questions.size();
		Scanner myObj = new Scanner(System.in);
		int i = 0;
		String takeAgain = "Y";
		do {

			if (questions.get(i).correct == false) {
				System.out.print("-------------------------\n");
				questions.get(i).showQuestion();
				String userGuess = myObj.nextLine();

				if (questions.get(i).checkAnswer(userGuess)) {
					System.out.println("Correct! Good Job!");
					questions.get(i).correct = true;
					correct++;

				} else {
					System.out.print("Sorry the Answer was ");
					questions.get(i).showAnswer();
				}
			}
			i++;

			if (correct == totalUnErrorQuestions) {
				return correct;
			}
			if (i == size) {
				i = 0;
				System.out.print("Do you want to retry the one's you missed? Y/N?");
				takeAgain = myObj.nextLine();
			}
		} while (takeAgain.charAt(0) == 'Y' || takeAgain.charAt(0) == 'y');

		return correct;
	}

//-------------------------------------------------------
	/**
	 * This function returns the total amount of questions in the quiz and returns
	 * it so we can find the percentage the user got on the quiz.
	 * 
	 * @return int Returns the num of total questions in the quiz
	 */
	static int numQuestions() {
		int capacity = questions.size();
		return capacity;
	}
}

//-------------------------------------------------------
